lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -420,6 +420,63 @@ void SPIRVRegularizeLLVMBase::cleanupConversionToNonStdIntegers(Module *M) {
   }
 }
 
+/// Move SPIRV builtin global variables to local memory
+/// Example: __spirv_BuiltInWorkgroupId is the builtin GV
+/// @__spirv_BuiltInWorkgroupId = external dso_local local_unnamed_addr
+///                               addrspace(1) constant <3 x i64>, align 32
+/// define weak_odr dso_local spir_kernel void @foo() {
+/// ...
+/// %0 = load i64, ptr addrspace(1) @__spirv_BuiltInWorkgroupId, align 32
+/// ret void
+/// }
+/// Transformed to --->
+/// @__spirv_BuiltInWorkgroupId = external dso_local local_unnamed_addr
+///                               addrspace(1) constant <3 x i64>, align 32
+/// define weak_odr dso_local spir_kernel void @foo() {
+/// ...
+/// %1 = load <3 x i64>, ptr addrspace(1) @__spirv_BuiltInWorkgroupId, align 32
+/// %__local_SPIRV_Builtin* = alloca <3 x i64> addrspace(0), align 32
+/// store <3 x i64> %1, ptr addrspace(0) %__local_SPIRV_Builtin*, align 32
+/// // Replace uses of __spirv_BuiltInWorkgroupId with %__local_SPIRV_Builtin*
+/// ret void
+/// }
+void SPIRVRegularizeLLVMBase::copyBuiltinGVToLocalVar(Module *M) {
+  for (auto I = M->begin(), E = M->end(); I != E;) {
+    Function *F = &(*I++);
+    for (BasicBlock &BB : *F) {
+      for (Instruction &Inst : BB) {
+        for (auto &Op : Inst.operands()) {
+          if (GlobalVariable *GV =
+                  dyn_cast<GlobalVariable>(Op->stripPointerCasts())) {
+            spv::BuiltIn Builtin = spv::BuiltInPosition;
+            if (!(GV->hasName() &&
+                  getSPIRVBuiltin(GV->getName().str(), Builtin)))
+              continue;
+            IRBuilder<> Builder(&Inst);
+            auto *LoadTy = GV->getValueType();
+            auto *LoadedVal = Builder.CreateLoad(LoadTy, GV);
+            if (GV->getAlignment())
+              LoadedVal->setAlignment(llvm::Align(GV->getAlignment()));
+            auto *NewVar = Builder.CreateAlloca(GV->getType(), nullptr,
```

Comment:
We might be having several builtins accessed in the kernel (global id, local id, size etc), if I read the code correctly it would result in several __local_SPIRV_Builtin declarations thus SSA violation. Regardless how it's addressed, please reflect it in the test.